src/vdom/component.js

```diff
@@ -19,8 +19,8 @@ export function setComponentProps(component, props, renderMode, context, mountAl
 	if (component._disable) return;
 	component._disable = true;
 
-	if ((component.__ref = props.ref)) delete props.ref;
-	if ((component.__key = props.key)) delete props.key;
+	if ((component.__ref = props.ref)!=null) delete props.ref;
```

Comment:
I don't get how this is suppose to work when props.ref is undefined.

Tried it in node cli
```js
> props = {}
{}
> props.ref = undefined
undefined
> component = {}
{}
> if((component.__ref = props.ref) != null) delete props.ref
undefined
> props
{ ref: undefined }
```

Reply:
@developit

Reply:
Thanks for pointing it out, @Pyrolistical. Have added an additional check for `undefined` as well. Not sure if it is optimal enough. @developit knows more about it, from my understanding.

Reply:
A check for `!= undefined` won't help.
The problem occurs if ref *is* `undefined`.
Also in javascript a condition with "normal" equality operators aren't type bound.
That means `undefined == null` is `true`.
Which means the second `!= undefined` is completly pointless.
For more informations on equality in js you could read the [MDN article](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators).

Reply:
Not exactly ideal, but this would work, handling either `null` or `undef`:

```js
if (!!(component.__ref = props.ref)) delete props.ref;
```

Reply:
`!!` won't work either.

Reply:
How about this?

```
props = {};
props.ref = undefined;
component = {};
console.log("before");
console.log({ component });
console.log({ props });
if ((component.__ref = props.ref && !!props.ref)|| props.ref==undefined || props.ref==null) delete props.ref;
console.log("after");
console.log({ component });
console.log({ props });
```
gives the output
```
node v9.7.1 linux/amd64
before
{ component: {} }
{ props: { ref: undefined } }
after
{ component: { __ref: undefined } }
{ props: {} }
=> undefined
```

It works similarly for `ref == null` as well.

Source: https://repl.it/repls/CalmFrivolousScience

Reply:
Works, but in that case the condition is completly pointless because it will *always* resolve to `true`.
No matter if `ref` is defined or not.
That means in that case it we could just omit the codition and roll back to a straight `delete props.ref;`

I'd need to profile if the performance gain would be worth the extra size when changing the condition to `('ref' in props) delete props.ref;` instead of a straight `delete props.ref;

Reply:
Thanks for the inputs, @Kanaye. In my humble opinion, the condition will _always_ resolve to `true` because `ref` must be deleted from `props` before further processing.